README.md
# Traversal

A service for hash map (assoc array) traversal.

When dealing with nested associative structures, traversing them can become
quite a pain. Mostly because of the amount of `isset` checking that is
necessary.

For example, to access a nested key `['foo']['bar']['baz']`, you must do
something like this:

```php
$baz = (isset($data['foo']['bar']['baz'])) ? $data['foo']['bar']['baz'] : null;
```

Enough already! `Traversal` provides a better way:

```php
$baz = Gnugat\Traversal\get_in($data, array('foo', 'bar', 'baz'));
```

> **Note**: `Traversal` is a fork of [Igor](https://igor.io/)'s [get-in](https://github.com/igorw/get-in).
> which provides support for PHP 5.3.

[![SensioLabsInsight](https://insight.sensiolabs.com/projects/31a73802-2b87-4d65-9ae1-52edaac9f2a6/mini.png)](https://insight.sensiolabs.com/projects/31a73802-2b87-4d65-9ae1-52edaac9f2a6)
[![Travis CI](https://travis-ci.org/gnugat/traversal.png)](https://travis-ci.org/gnugat/traversal)

## Installation

Use [Composer](http://getcomposer.org/) to install Traversal in your projects:

    composer require gnugat/traversal:~2.0

## Usage

Make the functions available by requiring Composer's autoloader:

```php
<?php

require __DIR__.'/vendor/autoload.php';
```

### Get in

Retrieve value from a nested structure using a list of keys:

```php
$users = array(
    array('name' => 'Igor Wiedler'),
    array('name' => 'Jane Doe'),
    array('name' => 'Acme Inc'),
);

$name = Gnugat\Traversal\get_in($users, array(1, 'name'));
//= 'Jane Doe'
```

Non existent keys return null:

```php
$data = array('foo' => 'bar'];

$baz = Gnugat\Traversal\get_in($data, array('baz'));
//= null
```
You can provide a default value that will be used instead of null:

```php
$data = array('foo' => 'bar');

$baz = Gnugat\Traversal\get_in($data, array('baz'), 'qux');
//= 'qux'
```

### Update in

Apply a function to the value at a particular location in a nested structure:

```php
$data = array('foo' => array('answer' => 42));
$inc = function ($x) {
    return $x + 1;
};

$new = Gnugat\Traversal\update_in($data, array('foo', 'answer'), $inc);
//= array('foo' => array('answer' => 43))
```

You can variadically provide additional arguments for the function:

```php
$data = array('foo' => 'bar');
$concat = function (/* $args... */) {
    return implode('', func_get_args());
};

$new = Gnugat\Traversal\update_in($data, array('foo'), $concat, ' is the ', 'best');
//= array('foo' => 'bar is the best')
```

### Assoc in

Set a value at a particular location:

```php
$data = array('foo' => 'bar');

$new = Gnugat\Traversal\assoc_in($data, array('foo'), 'baz');
//= array('foo' => 'baz')
```

It will also set the value if it does not exist yet:

```php
$data = [];

$new = Gnugat\Traversal\assoc_in($data, array('foo', 'bar'), 'baz');
//= array('foo' => array('bar' => 'baz'))
```

## Inspiration

The naming and implementation is inspired by the `get-in`, `update-in` and
`assoc-in` functions from [clojure](http://clojure.org).


CHANGELOG.md
# CHANGELOG

## 2.0.0

* removed all short array syntax
* moved Traversal methods into standalone functions (BC Break)

## 1.0.1

* removed PHP 5.4 callable type hint

## 1.0.0

* used PHP 5.3 array syntax
* encapsulated get-in functions in Traversal service
* fixed igorw/get-in issue #13


tests/Gnugat/Traversal/TraversalTest.php
<?php

/*
 * This file is part of the Traversal project.
 *
 * (c) Igor Wiedler <igor@wiedler.ch>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace tests\Gnugat\Traversal;

use PHPUnit_Framework_TestCase;

class TraversalTest extends PHPUnit_Framework_TestCase
{
    /** @dataProvider provideGetIn */
    public function testGetIn($expected, $array, $keys, $default = null)
    {
        $this->assertSame($expected, \Gnugat\Traversal\get_in($array, $keys, $default));
    }

    public function provideGetIn()
    {
        $single = array('key' => 'value');
        $nested = array('foo' => array('bar' => array('baz' => 'value')));
        $list   = array(array('name' => 'foo'));

        return array(
            array('value', $single, array('key'), 'default'),
            array(array('bar' => array('baz' => 'value')), $nested, array('foo'), 'default'),
            array(array('baz' => 'value'), $nested, array('foo', 'bar'), 'default'),
            array('value', $nested, array('foo', 'bar', 'baz'), 'default'),
            array('default', $nested, array('foo', 'bar', 'bang'), 'default'),
            array('default', $nested, array('non_existent'), 'default'),
            array(null, $nested, array('non_existent')),
            array($nested, $nested, array(), 'default'),
            array($nested, $nested, array()),
            array('foo', $list, array(0, 'name')),
            array(null, array('foo' => null), array('foo'), 'err'),
        );
    }

    /** @dataProvider provideUpdateIn */
    public function testUpdateIn($expected, $array, $keys, $fn, array $args = array())
    {
        $this->assertSame($expected, call_user_func_array(
            'Gnugat\Traversal\update_in',
            array_merge(array($array, $keys, $fn), $args)
        ));
    }

    public function provideUpdateIn()
    {
        $nested = array('foo' => array('bar' => array('baz' => 40)));
        $single = array('key' => 'value');

        $add = function ($a, $b) { return $a + $b; };
        $identity = function ($x) { return $x; };

        return array(
            array(array('foo' => array('bar' => array('baz' => 42))), $nested, array('foo', 'bar', 'baz'), $add, array(2)),
            array(array('foo' => array('bar' => array('baz' => 40))), $nested, array('foo', 'bar', 'baz'), $identity),
            array(array('foo' => array('bar' => array('baz' => 40))), $nested, array(), $identity),
            array(array('key' => 'value'), $single, array(), $identity),
            array(array('foo' => null), array('foo' => null), array('foo'), $identity),
        );
    }

    /**
     * @dataProvider provideInvalidUpdateIn
     * @expectedException InvalidArgumentException
     */
    public function testInvalidUpdateIn($expected, $array, $keys, $fn, array $args = array())
    {
        $this->assertSame($expected, call_user_func_array(
            'Gnugat\Traversal\update_in',
            array_merge(array($array, $keys, $fn), $args)
        ));
    }

    public function provideInvalidUpdateIn()
    {
        $nested = array('foo' => array('bar' => array('baz' => 40)));

        $add = function ($a, $b) { return $a + $b; };
        $identity = function ($x) { return $x; };

        return array(
            array(array('foo' => array('bar' => array('baz' => 40))), $nested, array('non_existent'), $identity),
            array(array('foo' => array('bar' => array('baz' => 40))), $nested, array('non', 'existent'), $identity),
        );
    }

    /** @dataProvider provideAssocIn */
    public function testAssocIn($expected, $array, $keys, $value)
    {
        $this->assertSame($expected, \Gnugat\Traversal\assoc_in($array, $keys, $value));
    }

    public function provideAssocIn()
    {
        $nested = array('foo' => array('bar' => array('baz' => 'value')));
        $single = array('key' => 'value');
        $empty  = array();

        return array(
            array(array('foo' => array('bar' => array('baz' => 'new value'))), $nested, array('foo', 'bar', 'baz'), 'new value'),
            array(array('key' => 'value'), $single, array(), 'new value'),
            array(array('foo' => array('bar' => 'new value')), $empty, array('foo', 'bar'), 'new value'),
            array(array('foo' => 'new value'), array('foo' => null), array('foo'), 'new value'),
            array(array('foo' => array('bar' => 'new value')), array('foo' => null), array('foo', 'bar'), 'new value'),
        );
    }
}


src/Gnugat/Traversal/get_in.php
<?php

/*
 * This file is part of the Traversal project.
 *
 * (c) Igor Wiedler <igor@wiedler.ch>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Gnugat\Traversal;

/**
 * Retrieve value from a nested structure using a list of keys:
 *
 * ```php
 * $users = array(
 *     array('name' => 'Igor Wiedler'),
 *     array('name' => 'Jane Doe'),
 *     array('name' => 'Acme Inc'),
 * );
 *
 * $name = Gnugat\Traversal\get_in($users, array(1, 'name'));
 * //= 'Jane Doe'
 * ```
 *
 * Non existent keys return null:
 *
 * ```php
 * $data = array('foo' => 'bar'];
 *
 * $baz = Gnugat\Traversal\get_in($data, array('baz'));
 * //= null
 * ```
 * You can provide a default value that will be used instead of null:
 *
 * ```php
 * $data = array('foo' => 'bar');
 *
 * $baz = Gnugat\Traversal\get_in($data, array('baz'), 'qux');
 * //= 'qux'
 * ```
 *
 * @param array $array
 * @param array $keys
 * @param mixed $default
 *
 * @return mixed
 */
function get_in(array $array, array $keys, $default = null) {
    if (!$keys) {
        return $array;
    }
    // This is a micro-optimization, it is fast for non-nested keys, but fails for null values
    if (count($keys) === 1 && isset($array[$keys[0]])) {
        return $array[$keys[0]];
    }
    $current = $array;
    foreach ($keys as $key) {
        if ($current === null || !array_key_exists($key, $current)) {
            return $default;
        }
        $current = $current[$key];
    }

    return $current;
}


src/Gnugat/Traversal/assoc_in.php
<?php

/*
 * This file is part of the Traversal project.
 *
 * (c) Igor Wiedler <igor@wiedler.ch>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Gnugat\Traversal;

/**
 * Set a value at a particular location:
 *
 * ```php
 * $data = array('foo' => 'bar');
 *
 * $new = Gnugat\Traversal\assoc_in($data, array('foo'), 'baz');
 * //= array('foo' => 'baz')
 * ```
 *
 * It will also set the value if it does not exist yet:
 *
 * ```php
 * $data = [];
 *
 * $new = Gnugat\Traversal\assoc_in($data, array('foo', 'bar'), 'baz');
 * //= array('foo' => array('bar' => 'baz'))
 * ```
 *
 * @param array $array
 * @param array $keys
 * @param mixed $value
 *
 * @return array
 */
function assoc_in(array $array, array $keys, $value) {
    if (!$keys) {
        return $array;
    }
    $current = &$array;
    foreach ($keys as $key) {
        if ($current === null || !is_array($current)) {
            $current = array();
        }
        $current = &$current[$key];
    }
    $current = $value;

    return $array;
}


src/Gnugat/Traversal/update_in.php
<?php

/*
 * This file is part of the Traversal project.
 *
 * (c) Igor Wiedler <igor@wiedler.ch>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Gnugat\Traversal;

/**
 * Apply a function to the value at a particular location in a nested structure:
 *
 * ```php
 * $data = array('foo' => array('answer' => 42));
 * $inc = function ($x) {
 *     return $x + 1;
 * };
 *
 * $new = Gnugat\Traversal\update_in($data, array('foo', 'answer'), $inc);
 * //= array('foo' => array('answer' => 43))
 * ```
 *
 * You can variadically provide additional arguments for the function:
 *
 * ```php
 * $data = array('foo' => 'bar');
 * $concat = function () {
 *     return implode('', func_get_args());
 * };
 *
 * $new = Gnugat\Traversal\update_in($data, array('foo'), $concat, ' is the ', 'best');
 * //= array('foo' => 'bar is the best')
 * ```
 *
 * @param array    $array
 * @param array    $keys
 * @param callable $f
 *
 * @return array
 */
function update_in(array $array, array $keys, $f /* , $args... */) {
    $args = array_slice(func_get_args(), 3);
    if (!$keys) {
        return $array;
    }
    $current = &$array;
    foreach ($keys as $key) {
        if ($current === null || !array_key_exists($key, $current)) {
            throw new \InvalidArgumentException(sprintf('Did not find path %s in structure %s', json_encode($keys), json_encode($array)));
        }
        $current = &$current[$key];
    }
    $current = call_user_func_array($f, array_merge(array($current), $args));

    return $array;
}
